lib/string_basics.dart

```diff
@@ -200,6 +202,41 @@ extension StringBasics on String {
     }
     return stringBuffer.toString();
   }
+
+  /// Returns a truncated version of the string.
+  ///
+  /// Example:
+  /// ```dart
+  /// final sentence = 'The quick brown fox jumps over the lazy dog';
+  /// final truncated = sentence.truncate(20); // 'The quick brown fox...'
+  /// ```
+  ///
+  /// The [length] is the truncated length of the string.
+  /// The [substitution] is the substituting string of the truncated characters.
+  /// If not null or empty it will be appended at the end of the truncated string.
+  /// The [trimTrailingWhitespace] is whether or not to trim the spaces of the truncated string
+  /// before appending the ending string.
+  /// The [includeSubstitutionInLength] is whether or not that the length of the substitution string will be included
+  /// with the intended truncated length.
+  String truncate(
+    int length, {
+    String substitution = '',
+    bool trimTrailingWhitespace = true,
+    bool includeSubstitutionInLength = false,
+  }) {
+    if (this.length <= length) {
+      return this;
+    }
+
+    // calculate the final truncate length where whether or not to include the length of substitution string
+    final truncatedLength =
+        includeSubstitutionInLength ? (length - substitution.length) : length;
```

Comment:
`substitution.length` returns the number of UTF-16 code units, not the number of human-perceived characters (grapheme clusters). This can be fixed by using `substitution.characters.length` instead. You can add `'abc'.truncate(2, substitution: '👍', includeSubstitutionInLength: true)` as a test: It returns `'👍'` instead of `'a👍'`.

Reply:
Thanks! you're right. I added a test and updated the `substitution.characters.length` extension.